pkg/reconciler/accessor/networking/certificate.go

```diff
@@ -0,0 +1,87 @@
+limitations under the License.
+*/
+
+package networking
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/logging"
+	"knative.dev/serving/pkg/apis/networking/v1alpha1"
+	clientset "knative.dev/serving/pkg/client/clientset/versioned"
+	listers "knative.dev/serving/pkg/client/listers/networking/v1alpha1"
+	kaccessor "knative.dev/serving/pkg/reconciler/accessor"
+)
+
+// CertificateAccessor is an interface for accessing Knative Certificate.
+type CertificateAccessor interface {
+	GetServingClient() clientset.Interface
+	GetCertificateLister() listers.CertificateLister
+}
+
+// ReconcileCertificate reconciles Certificate to the desired status.
+func ReconcileCertificate(ctx context.Context, owner kmeta.Accessor, desired *v1alpha1.Certificate,
+	certAccessor CertificateAccessor) (*v1alpha1.Certificate, error) {
+
+	logger := logging.FromContext(ctx)
+	recorder := controller.GetEventRecorder(ctx)
+	if recorder == nil {
+		return nil, fmt.Errorf("recoder for reconciling Certificate %s/%s is not created", desired.Namespace, desired.Name)
+	}
+	cert, err := certAccessor.GetCertificateLister().Certificates(desired.Namespace).Get(desired.Name)
+	if apierrs.IsNotFound(err) {
+		cert, err = certAccessor.GetServingClient().NetworkingV1alpha1().Certificates(desired.Namespace).Create(desired)
+		if err != nil {
+			logger.Error("Failed to create Certificate", zap.Error(err))
+			recorder.Eventf(owner, corev1.EventTypeWarning, "CreationFailed",
+				"Failed to create Certificate %s/%s: %v", desired.Namespace, desired.Name, err)
+			return nil, err
+		}
+		recorder.Eventf(owner, corev1.EventTypeNormal, "Created", "Created Certificate %s/%s", cert.Namespace, cert.Name)
+		return cert, nil
+	} else if err != nil {
+		return nil, err
+	} else if !metav1.IsControlledBy(cert, owner) {
+		// Return an error with NotControlledBy information.
+		return nil, kaccessor.NewAccessorError(
+			fmt.Errorf("owner: %s with Type %T does not own Certificate: %q", owner.GetName(), owner, cert.Name),
+			kaccessor.NotOwnResource)
+	} else {
+		if !equality.Semantic.DeepEqual(cert.Spec, desired.Spec) {
+			// Don't modify the informers copy
+			existing := cert.DeepCopy()
+			existing.Spec = desired.Spec
+			cert, err := certAccessor.GetServingClient().NetworkingV1alpha1().Certificates(existing.Namespace).Update(existing)
+			if err != nil {
+				recorder.Eventf(owner, corev1.EventTypeWarning, "UpdateFailed",
+					"Failed to update Certificate %s/%s: %v", existing.Namespace, existing.Name, err)
+				return nil, err
+			}
+			recorder.Eventf(owner, corev1.EventTypeNormal, "Updated",
+				"Updated Spec for Certificate %s/%s", existing.Namespace, existing.Name)
+			return cert, nil
```

Comment:
remove this return path, since the same code will execute below.

Reply:
Done